Fix #17 by only looking at un-selected tiles

mattnield opened the pull request:
Change so that we only look at the tiles that have not already been drawn. Also starts to put some things in place to display alternative tiles in the grid for drawn cards - one for winners, and one for no-shows.

mattnield committed:
Fix #17 by only selecting tiles from those that have not already been drawn and start working on replacing the profile photo images for drawn tiles

dracan commented on PrizeDraw/MainWindow.xaml.cs:
```diff
@@ -18,6 +19,8 @@ public partial class MainWindow : Window
         private const int WinnerTileTargetHeight = 500;
         private readonly MainWindowViewModel _viewModel;
+        private static string AppFolder => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "PrizeDraw");
+        private static string imageFolder => Path.Combine(AppFolder, "Images");
```
Can `imageFolder` be capitalised? Also, why are these images being stored in the AppData folder? This folder won't exist until you run the app for the first time. These images should really be with the application.

mattnield replied:
I was just reusing the same code used to load the profile images and the 'default' image that we've been using (unless I've fundamentally misunderstood that code). I'll take a look at moving them back in with the application.

I can capitalise `ImageFolder`, I've always tended to use camel-case for private, but happy to change that.

dracan replied:
R# (and I'm assuming VS) suggest that static properties should be upper camel-case. For non-static, it would be camel-case with an underscore prefix. I really need to add an editorconfig file to the project.
However, _both_ of `AppFolder` and `imageFolder` are redundant, and don't do anything - so both can be removed.

dracan replied:
> I was just reusing the same code used to load the profile images and the 'default' image that we've been using (unless I've fundamentally misunderstood that code)

Ah, I was misreading it because both of your `AppFolder` and `imageFolder` are redundant here - and aren't used. Ie. when you had the unused `imageFolder` being passed into `LoadWinnerImage` and `LoadNoshowImage` - it looked liked your winner and noshow images were coming from the AppFolder (which wouldn't be right). But because you weren't using the `imageFolder` parameter, it's fine - your `@"Images/noshow.png"` will be looking in the exe folder, as it should do.

dracan replied:
So, TL;DR version is just to delete the unused `AppFolder` property, and all's good :)

dracan commented on PrizeDraw/MainWindow.xaml.cs:
```diff
@@ -127,61 +135,73 @@ async void MainWindow_KeyDown(object sender, KeyEventArgs e)
+                            _viewModel.SelectedTile.LoadNoshowImage(imageFolder);
+                        }
+
+                        // Can we replace the tile with the VieUserControl
```
Currently, the images are just loaded on initialisation. However, if we make the list of tiles an `ObservableCollection`, we can make it more dynamic. I'd have to have a play around with it though to get this to work. I'm happy to have a look if you'd like? But it would have to be after my talk on Monday - I need to focus on that first.

mattnield replied:
Sounds like a reasonable enhancement. My motivation for the PR really is to prevent us from drawing people more than once per event (whether a winner or a no-show). I think perhaps add a new issue for that to keep things progressing?

mattnield replied:
Looking at the changes the I've just committed and other comments, the image location is no longer an issue. Just the `ObservableCollection` to look at.

dracan commented on PrizeDraw/ViewModels/TileViewModel.cs:
```diff
@@ -28,12 +42,33 @@ public void LoadImage(string imageFolder)
+            BitmapImage = LoadBitmap(image);
+        }
+
+        public void LoadWinnerImage(string imageFolder)
```
`imageFolder` isn't being used.

dracan commented on PrizeDraw/ViewModels/TileViewModel.cs:
```diff
@@ -28,12 +42,33 @@ public void LoadImage(string imageFolder)
+            BitmapImage = LoadBitmap(image);
+        }
+
+        public void LoadNoshowImage(string imageFolder)
```
`imageFolder` isn't being used.

dracan commented on PrizeDraw/ViewModels/TileViewModel.cs:
```diff
@@ -14,6 +14,20 @@ public class TileViewModel : ViewModelBase
+            set
+            {
+                _isWinner = value;
+                _isDrawn = true;
```
Can we not do this :) As a user of this property, I wouldn't expect setting `IsWinner` to have a magical side effect of also setting `IsDrawn`. Can we change this to be just normally properties, and set both explicitly in the calling code?

mattnield replied:
Why not? Surely it's implicit. If you've won the draw then you've been drawn. Don't get me wrong, I'm not precious about it but you can't win without being drawn so it seemed logical to set them at the same time to me.

dracan replied:
It's quite dangerous code and could well introduce future bugs. Imagine requirements change, and someone's working on some code, and writes the following...

    tile.IsDrawn = false;
    tile.IsWinner = true;

Those two lines are very simple, and one would expect the value of `IsDrawn` to be false. But with your change, `IsDrawn` would be _magically_ set to true. As a user of these properties, I wouldn't expect setting a simple property called `IsWinner` to be changing other properties behind the scenes. To know that this was happening, I'd have had to have viewed the definition of that property.

Hidden side effects are almost always a bad thing - both properties and methods should do what they say on the tin, and try to avoid hidden other stateful side effects.

dracan replied:
In your new commit, can't you just use auto-properties? Ie. you've currently got all this...

![image](https://user-images.githubusercontent.com/567988/68081797-6f7bf200-fe0b-11e9-83e2-4cf2bfeef42a.png)

Rather than just:

    public bool IsWinner { get; set; }
    public bool IsDrawn { get; set; }

dracan commented on PrizeDraw/ViewModels/MainWindowViewModel.cs:
```diff
@@ -159,13 +161,11 @@ private void HandleTimer()
             }
 
             var rand = new Random();
+            var availableTiles = Tiles.ToList().Where(t => !t.IsDrawn).ToList();
```
The first `ToList()` is redundant.

dracan commented on PrizeDraw/ViewModels/MainWindowViewModel.cs:
```diff
@@ -144,6 +144,8 @@ public void SaveWinnerDetails(TileViewModel winningTileViewModel, bool flagged)
 
         private void HandleTimer()
         {
+            // to randomise the tiles. Any tile drawn has the attendee id removed. pressing Enter
```
Pedantic, I know - but can the sentence be capitalised? Also, most of this comment seems unrelated to the method it's in?

mattnield replied:
Ah - was my first `TODO` comment, I'll remove this now as it's no longer needed.

dracan commented:
Hey @mattnield. I've added a few small comments. Liking the look of those new icons - will be interested to see how it looks when this is working :) If you need a hand getting the databinding to work, let me know. But will have to be after my talk on Monday, as I need to focus on that. I've left it a bit late to finish off my talk! Oops! ;)

mattnield commented:
No worries, @dracan - focus on the meetup :) 

I'll catch up with you at some point soon hopefully to maybe go through the comments (I can't actually see any just now, which is odd). It'd be good to look at how the binding works and also at how the tiles toggle when the shuffle happens etc., as I couldn't find anything obvious there.

dracan commented:
Oops! I didn't realise I had to submit the review! 🤣 I'm used to ADOS where when you add a comment to a PR, the other person sees it straight away.

mattnield committed:
Ensure that `IsWinner` and `IsDrawn` are not linked via the `IsWinnder` setter

mattnield committed:
Remove redundant comment

mattnield committed:
Remove redundant variable used to load images

mattnield commented:
Resolved most of those - no need to look at them now- just wanted to get them done.

dracan committed:
Made the winner and noshow images work

Instead of hijacking the users' image for this, I've added a new XAML block to the TileUserControl for the noshow tile and the winner tile. Then I'm using flags to toggle the visibility.

dracan committed:
Change winner and noshow images

The other ones were quite blurry and had a grey background

dracan commented:
@mattnield I've pushed some changes which gets this to work. Rather than trying to replace the users' profile pic with the no-show/winner images - I've added a new block to the TileUserControl XAML, and it uses databinding to determine which block is visible. This seems to work quite well.

Unfortunatley, your images were quite blurry when it was working - so I've changed them - hope that's okay? It looks like this now...

![image](https://user-images.githubusercontent.com/567988/68547070-0f6de880-03d5-11ea-8357-0ebe4ce27b94.png)

dracan committed:
Reduce whitespace around winner image

This makes it closer to the size of the no-show image.
